### ISSUE TITLE
Exception when calling ListManagedInstances and other methods with string parameters on a pageless managed instance group

### ISSUE THREAD

**Issue by matt-edmondson** [googleapis/google-cloud-dotnet#13311]
#### Environment details

  - OS: Windows 11
  - .NET version: 8.0
  - Package name and version: Google.Cloud.Compute.V1 2.16.0

#### Steps to reproduce

  1. Call `ListManagedInstances` on a pageless managed instance group via the overload that accepts string parameters
  2. Observe the Grpc.Core.RpcException: StatusCode="InvalidArgument", Detail="Invalid value for field 'pageToken': ''. Supplied restart token corresponds to a zone not supported by this managed instance group."
 
#### Explanation
In the string parameter overloads for this and several other methods in the API the `pageToken` property of the request is incorrectly null-coalesced to an empty string.

When using the overload that accepts a request object the `pageToken` is initialized to `null` which is the correct value that the API is expecting and the call succeeds.

#### Rectification
Either remove the null-coalesces in this and other method overloads, or make the API tolerant of the empty string in this property.

**Comment by jskeet** [comment]
I've managed to reproduce this, and I'll now work out the best way of addressing it.

To give a bit of background for what I think is going on... historically (when the generator was first written) this wasn't a problem - or rather, using an empty string was the right thing to do - because it predated proto3 `optional` fields. The `ListManagedInstancesInstanceGroupManagersRequest.PageToken` property will never actually be null (and can't be assigned a null value) - but proto3 `optional` fields support *field presence* so there's a difference between *explicitly* setting it to an empty string and leaving it with the default value without setting it.

For example, this code throws an `ArgumentNullException`:

```csharp
var request = new ListManagedInstancesInstanceGroupManagersRequest
{
    PageToken = null
};
```

... whereas this prints 0:

```csharp
Console.WriteLine(new ListManagedInstancesInstanceGroupManagersRequest().PageToken.Length);
```

The Compute API is different to most APIs we generate "GAPIC" libraries from. While we generate it from a proto, that proto is actually itself generated from Discovery docs. This explains why there are so many more optional fields in the Compute API than in most others - indeed, I can only find *one* other API that has a proto3-optional `page_token` field.

That said, this appears to be an inconsistency even *within* Compute - we have integration tests using `AcceleratorTypesClient` for example, listing only with the constituent parts. I suspect it's the unusual aspect of "pageless vs paged" that is affecting things here. I'll raise this with the Compute team and see if they'd be happy to ignore explicit empty page tokens. (Arguably we *also* shouldn't be setting `max_results` when it's not provided.)

In the meantime, I'll look at what we might want to generate instead of our current code. Thanks very much for the bug report, and I hope this slightly rambling explanation was useful.

**Comment by jskeet** [comment]
Right - this is the new code I would like to generate for `ListManagedInstances`:

```csharp
public virtual gax::PagedEnumerable<InstanceGroupManagersListManagedInstancesResponse, ManagedInstance> ListManagedInstances(string project, string zone, string instanceGroupManager, string pageToken = null, int? pageSize = null, gaxgrpc::CallSettings callSettings = null)
{
    var request = new ListManagedInstancesInstanceGroupManagersRequest
    {
        InstanceGroupManager = gax::GaxPreconditions.CheckNotNullOrEmpty(instanceGroupManager, nameof(instanceGroupManager)),
        Project = gax::GaxPreconditions.CheckNotNullOrEmpty(project, nameof(project)),
        Zone = gax::GaxPreconditions.CheckNotNullOrEmpty(zone, nameof(zone)),
    };
    if (pageToken is not null)
    {
        request.PageToken = pageToken;
    }
    if (pageSize is not null)
    {
        request.PageSize = pageSize.Value;
    }
    return ListManagedInstances(request, callSettings);
}
```

We should be able to generate code like that (both sync and async) regardless of the "optionality" of the field, so we don't need any conditional code in the generator. I'll start work on the generator change now.

**Comment by jskeet** [comment]
Alternative generation approach, using a new extension method (on `IPageRequest`) we could add in GAX:

```csharp
public virtual gax::PagedEnumerable<InstanceGroupManagersListManagedInstancesResponse, ManagedInstance> ListManagedInstances(string project, string zone, string instanceGroupManager, string pageToken = null, int? pageSize = null, gaxgrpc::CallSettings callSettings = null) =>
ListManagedInstances(new ListManagedInstancesInstanceGroupManagersRequest
            {
                InstanceGroupManager = gax::GaxPreconditions.CheckNotNullOrEmpty(instanceGroupManager, nameof(instanceGroupManager)),
                Project = gax::GaxPreconditions.CheckNotNullOrEmpty(project, nameof(project)),
                Zone = gax::GaxPreconditions.CheckNotNullOrEmpty(zone, nameof(zone)),
            }.WithPaginationApplied(pageToken, pageSize), callSettings);
```

That would probably be somewhat simpler to implement in the generator (although admittedly it would need a new GAX release).

**Comment by jskeet** [comment]
Scratch that - I think we should be able to do the earlier version after all.

**Comment by jskeet** [comment]
I have a first pass at a fix, but was busy in meetings for the rest of the day. Will send out a PR for this tomorrow - I'm not promising anything (as getting a new generator version into production involves some hoops) but I'm reasonably hopeful that we can have a new Compute library published some time next week.

**Comment by jskeet** [comment]
(Reopening until we've got the change in this repo.)

**Comment by jskeet** [comment]
This has now been published in version 2.17.0. Please do shout if you still see any problems, and thanks again for the report.
